### PR TITLE
Feature/branded claims

### PR DESCRIPTION
**Pull request by nahu90**
**Issues:**
- [POAP-447](https://xivis.atlassian.net/browse/POAP-447)
- [POAP-438](https://xivis.atlassian.net/browse/POAP-438)
- [POAP-439](https://xivis.atlassian.net/browse/POAP-439)
- [POAP-440](https://xivis.atlassian.net/browse/POAP-440)

### CODE REVIEW

File: server/src/db/index.ts [hunk 1]
```diff
@@ -601,3 +638,87 @@ export async function getTotalQrClaims(eventId: number, qrRollId: number, claime
...
+export async function getPaginatedEventTemplates(limit: number, offset: number, name: string|null): Promise<EventTemplate[]> {
+  let query = `SELECT * FROM event_templates WHERE is_active = true `
+
+  if (name) {
+    query = query + `AND name ILIKE '${name}' `;
+  }
+
+  query = query + 'ORDER BY id DESC LIMIT ${limit} OFFSET ${offset}';
+
+  console.log(query);
```
Comment (ragonzal) [inline server/src/db/index.ts]: Please remove

File: server/src/db/index.ts [hunk 2]
```diff
@@ -601,3 +638,87 @@ export async function getTotalQrClaims(eventId: number, qrRollId: number, claime
...
+  return res
+}
+
+export async function getTotalEventTemplates(name: string): Promise<number> {
+  let query = `SELECT * FROM event_templates WHERE is_active = true `
+
+  if (name) {
+    query = query + `AND name ILIKE '${name}' `;
+  }
+  console.log(query);
```
Comment (ragonzal) [inline server/src/db/index.ts]: Please remove

File: server/src/db/index.ts [hunk 3]
```diff
@@ -28,6 +47,17 @@ export async function getEvents(): Promise<PoapEvent[]> {
...
+  const res = await db.manyOrNone<PoapEvent>('SELECT ' + publicEventColumns + ' FROM events ORDER BY start_date DESC');
+  return res.map(event => {
+    return {
+      ...event,
+      start_date: formatDate(event.start_date),
+      end_date: formatDate(event.end_date),
+    }
+  });
+}
+
```
Comment (ragonzal) [inline server/src/db/index.ts]: This function is never called, right? Please delete if not necessary

File: server/src/db/index.ts [hunk 4]
```diff
@@ -11,7 +30,7 @@ const db = pgPromise()({
 });
 
 const publicEventColumns = 'id, fancy_id, name, description, city, country, event_url, image_url, year, start_date, ' +
-  'end_date, event_host_id, from_admin, virtual_event'
+  'end_date, event_host_id, from_admin, virtual_event, event_template_id'
```
Comment (ragonzal) [inline server/src/db/index.ts:33]: Are we sure that altering this query does not changes the events API?

File: server/src/db/index.ts [hunk 5]
```diff
@@ -125,6 +155,7 @@ export async function getEvent(id: number | string): Promise<null | PoapEvent> {
       ...res,
       start_date: formatDate(res.start_date),
       end_date: formatDate(res.end_date),
+      event_template_id: res.event_template_id,
```
Comment (ragonzal) [inline server/src/db/index.ts]: If the `publicEventColumns` includes `event_template_id`, why are you including it?
Remember that `...res` means that is copying all elements from the object, and the overwriting (or creating) `start_date` and `end_date`

File: server/src/routes.ts [hunk 6]
```diff
@@ -535,6 +565,9 @@ export default async function routes(fastify: FastifyInstance) {
         return new createError.InternalServerError('Qr Claim does not have any event');
       }
       qr_claim.event = event;
+      if (event.event_template_id) {
+        qr_claim.event_template = await getEventTemplate(event.event_template_id);
+      }
```
Comment (ragonzal) [inline server/src/routes.ts:572]: Did you try this endpoint for QRs in different states and with/out template?

File: server/src/routes.ts [hunk 7]
```diff
@@ -2215,6 +2266,480 @@ export default async function routes(fastify: FastifyInstance) {
...
+      schema: {
+        description: 'List paginated event templates, you can filter by name',
+        tags: ['Event Templates',],
+        querystring: {
+          limit: { type: 'number' },
+          offset: { type: 'number' },
+          name: { type: 'string' },
+        },
+      },
+    },
```
Comment (ragonzal) [inline server/src/routes.ts]: Please add the response schema so that swagger can show this info please!

File: server/src/routes.ts [hunk 8]
```diff
@@ -2215,6 +2266,480 @@ export default async function routes(fastify: FastifyInstance) {
...
+                created_date: { type: 'string' },
+                is_active: { type: 'boolean' },
+              },
+            }
+          },
+          security: [
+            {
+              "authorization": []
+            }
+          ]
```
Comment (ragonzal) [inline server/src/routes.ts:2435]: `GET` endpoints should be public. Is this necessary?

File: server/src/routes.ts [hunk 9]
```diff
@@ -2215,6 +2266,480 @@ export default async function routes(fastify: FastifyInstance) {
...
+                main_color: { type: 'string' },
+                footer_color: { type: 'string' },
+                left_image_url: { type: 'string' },
+                left_image_link: { type: 'string' },
+                right_image_url: { type: 'string' },
+                right_image_link: { type: 'string' },
+                mobile_image_url: { type: 'string' },
+                mobile_image_link: { type: 'string' },
+                footer_icon: { type: 'string' },
+                secret_code: { type: 'string' },
```
Comment (ragonzal) [inline server/src/routes.ts]: `secret_code` should only be returned for admin. Where is that logic?

File: server/src/routes.ts [hunk 10]
```diff
@@ -2215,6 +2266,480 @@ export default async function routes(fastify: FastifyInstance) {
...
+              'main_color',
+              'footer_color',
+              'left_image_url',
+              'left_image_link',
+              'right_image_url',
+              'right_image_link',
+              'mobile_image_url',
+              'mobile_image_link',
+              'footer_icon',
+              'secret_code'
```
Comment (ragonzal) [inline server/src/routes.ts]: We missed talking about required fields. They are the following:
```
'name',
'title_image',
'title_link',
'header_color',
'header_link_color',
'main_color',
'footer_icon',
'secret_code'
```

Please review validations and refactor

File: server/src/routes.ts [hunk 11]
```diff
@@ -2215,6 +2266,480 @@ export default async function routes(fastify: FastifyInstance) {
...
+      async (req: any, res) => {
+        const title_image = req.body[Symbol.for('title_image')][0];
+        if (!title_image) {
+          return new createError.BadRequest('An image is required for field title_image');
+        }
+        if (title_image.mimetype != 'image/png') {
+          return new createError.BadRequest('title_image must be png');
+        }
+        const filename = 'title-image-' + (new Date().getTime()) + '.png'
+        const title_image_url = await uploadFile(filename, title_image.mimetype, title_image.data);
```
Comment (ragonzal) [inline server/src/routes.ts:2551]: I see that we are dropping the images in the Cloud Storage in the root folder. We talked about storing all of them in a folder.
I created in the bucket a folder named `templates/`. Please check if we can send all images to that folder please

**Review by ragonzal — CHANGES_REQUESTED** [review]
Please review all comments

File: server/src/db/index.ts (hunk 4 above)
Comment (nahu90) [reply server/src/db/index.ts:33]: Are you refering to the GET entpoint? it does not add event_template_id to GET endpoints if you dont add the field in response properties.

Screenshot from 2020-07-30 13-19-34

File: server/src/routes.ts (hunk 6 above)
Comment (nahu90) [reply server/src/routes.ts:572]: it works, but if the qr_claim event doesn't have any template you dont get the key 'event_template' in the response, in the next PR i'm adding a else to send the key in null.

File: server/src/routes.ts (hunk 8 above)
Comment (ragonzal) [reply server/src/routes.ts:2435]: @nahu90 in both the GET for all (`/`) and for a specific object (`/id`), we should return the secret code for the admins. So maybe the `security` is ok. But I don't se you handling the logic

**Review by ragonzal — APPROVED** [review]

**Review by ragonzal — APPROVED** [review]

**Pull request merged**